### ISSUE TITLE
About i-Mix for Supervised Contrastive Learning

### ISSUE THREAD

**Issue by quotation2520** [kibok90/imix#7]
Hello, thank you for the great work!

I have a question regarding the extension of i-Mix on supervised contrastive learning (Appendix section A.2, A.3 in your paper):

The supervised contrastive learning loss introduces the scaling term $N_{y_{i}}$ (the number of data with label $y_i$), and I'm not sure how to handle the number of positive samples for the mixed classes. 

Does the linearity of losses (proved in Appendix section B) hold in this case too? 
In other words, does the following equation hold for both SupCLR/ N-Pair SCL?

$l_{SupCon}(\lambda x_i + (1-\lambda)x_j, \lambda v_i + (1-\lambda) v_j ; B) = \lambda l_{SupCon}(\lambda x_i + (1-\lambda)x_j, v_i ; B) + (1-\lambda) l_{SupCon}(\lambda x_i + (1-\lambda)x_j, v_j ; B)$

If not, can you give the i-Mix implementation for supervised contrastive learning?
Thank you.

**Comment by kibok90** [comment]
For the supclr experiments in the appendix, I implemented i-mix on top of the official supclr code. They did not actually use the linearity of the loss for their implementation, but it also satisfies the linearity, as it can also be represented in the form of the cross-entropy loss. $N_{y_i}$ is the normalizer, which is the number of softmax probablities (each mixed sample is considered as one sample in i-mix) you added to compute the loss. Just make sure that you weight each (mixed) sample equally when computing the averaged loss.
https://github.com/HobbitLong/SupContrast
https://arxiv.org/pdf/2004.11362v1.pdf

**Comment by quotation2520** [comment]
Hi, thank you for the quick response. 

I'd like to make sure that I understand your work thoroughly. Please kindly answer the follow-up question

Let's say mixing ratio $\lambda = 0.7$, and there are three classes A, B, and C. (one-hot labels are (1 0 0), (0 1 0), (0 0 1), respectively.) 
Data mixing produces many new mixed classes, for instance  
M1: 0.7A + 0.3B   (0.7 0.3 0)
M2: 0.3A + 0.7B   (0.3 0.7 0)
M3: 0.7A + 0.3C   (0.7 0 0.3)

Are M1, M2, and M3 considered to be separate classes? Thank you for clarifying.

**Comment by kibok90** [comment]
I went through my old implementation and found that I generated two masks and then used the linearity to compute the final loss.
To answer your question, I think M1 is consider to be 0.7 of class A and 0.3 of class B, assuming that the math in the supcon paper is matched with their implementation in the official repo. If you want to be mathematically rigorous, you can do the math on your side to confirm it with the code below, which is the essential change from the original supcon repo. `mask2` is the mask for the samples with the multiplier $1-\lambda$.

https://github.com/HobbitLong/SupContrast/blob/master/losses.py#L92
`mean_log_prob_pos = (mask * log_prob).sum(1) / mask.sum(1)`
->
`mean_log_prob_pos = lam * (mask * log_prob).sum(1) / mask.sum(1) + (1. - lam) * (mask2 * log_prob).sum(1) / mask2.sum(1)`
